### PR TITLE
[in_app_purchase] Implementation of the app facing package

### PR DESCRIPTION
**Pull request by mvanbeusekom**
Implementation of the app facing package. This is the last part of the migration to the federated plugin architecture.

This PR is marked as draft and needs the following tasks to be completed:

- [x] Update dependencies to use pub.dev when the following packages have been published:
    - in_app_purchase_platform_interface: defines the interface that is to be implemented for different platforms (see PRs #3781, #3821, #3852);
    - in_app_purchase_android: the endorsed Android implementation (see PRs #3841, #3861) ;
    - in_app_purchase_ios: the endorsed iOS implementation (see PRs #3832, #3851, #3870) ;
- [x] Add missing breaking changes to the CHANGELOG.md;
- [x] Update the README.md to match the new implementation.

Related issues:
- Fixes flutter/flutter#78525
- Fixes flutter/flutter#79473
- Fixes flutter/flutter#60763
- Fixes flutter/flutter#45333
- Fixes flutter/flutter#45425

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format](../script/tool/README.md#format-code))
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,35 @@
+## 0.6.0
+
+As part of implementing federated architecture and making the interface compatible for other platform this version contains the following **breaking changes**:
+
+* Changes to the platform agnostic interface:
+  * The `InAppPurchaseConnection` class has been renamed to `InAppPurchase`. To access the generic InApp Purchase API use `InAppPurchase.instance`;
```
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase/CHANGELOG.md:6]: nits: maybe we can make this clearer such as


If you used `inAppPurchaseConnection.instance` to access generic In App Purchase APIs, please use `InAppPurchase.instance` instead.

File: packages/in_app_purchase/in_app_purchase/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+  * The `purchaseUpdatedStream` has been renamed to `purchaseStream`. Restored purchases are now also emitted on this stream and are marked with a `status` of `PurchaseStatus.restored`;
```
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase/CHANGELOG.md:7]: We don't have to mention "restored purchases are now also emitted on this stream and are marked with a `status` of `PurchaseStatus.restored`;"
Instead, we should give more details in the item below: 

The `InAppPurchaseConnection.queryPurchases` method has been removed. Instead, you should use `InAppPurchase.restorePurchases`. This method emits each restored purchase on the `InAppPurchase.purchaseStream`, the purchase object will be marked with a `status` of `PurchaseStatus.restored`

Also, for each API, mention its class, for example, `InAppPurchaseConnection.purchaseUpdatedStream`, and `InAppPurchase.purchaseStream`

Ditto below for all the other items.

File: packages/in_app_purchase/in_app_purchase_android/pubspec.yaml [hunk 3]
```diff
@@ -21,6 +21,7 @@ dependencies:
   flutter:
     sdk: flutter
 
+  json_annotation: ^4.0.1
```
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_android/pubspec.yaml:24]: Are the changes in in_app_purchase_android and in_app_purchase_ios here intentional?

**Review by cyanglaz — CHANGES_REQUESTED** [review]
Code mostly looks good. Also since there are API changes, don't we need to update README and example/README?

File: packages/in_app_purchase/in_app_purchase_android/pubspec.yaml (hunk 3 above)
Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase_android/pubspec.yaml]: Yes, they have been solved with PR #3879 which wasn't merge at that time. Now that it is merged and I rebased the changes have been removed from this PR.

**Comment by mvanbeusekom** [comment]
> Code mostly looks good. Also since there are API changes, don't we need to update README and example/README?

Yes that is why this PR is still in draft (also see checklist in the description)

File: packages/in_app_purchase/in_app_purchase/CHANGELOG.md (hunk 1 above)
Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase/CHANGELOG.md]: Done

File: packages/in_app_purchase/in_app_purchase/CHANGELOG.md (hunk 2 above)
Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase/CHANGELOG.md]: done

**Review by cyanglaz — APPROVED** [review]
LGTM

**Comment by mvanbeusekom** [comment]
Tree went red because the Xcode project (part of the `in_app_purchase` package) still referenced tests that had been moved to the `in_app_purchase_ios` package. 

I have removed these tests as they don't belong here.

**Pull request merged**